src/kafka-net/ManualConsumer.cs

```diff
@@ -0,0 +1,184 @@
+    {
+        private readonly string _topic;
+        private readonly int _partitionId;
+        private readonly ProtocolGateway _gateway;
+        private int _maxBytesInOneMessageIncludingHeader;
+        private readonly string _clientId;
+        private List<Message> _lastMessages;
+
+        private const int MaxWaitTimeForKafka = 0;
+        private const int UseBrokerTimestamp = -1;
+        private const int NoOffsetFound = -1;
+        private const double MessageSizeMultiplier = 1.5;
+
+        public ManualConsumer(int partitionId, string topic, ProtocolGateway gateway, string clientId)
+        {
+            if (string.IsNullOrEmpty(topic)) throw new ArgumentNullException("topic");
+            if (gateway == null) throw new ArgumentNullException("gateway");
+
+            _gateway = gateway;
+            _partitionId = partitionId;
+            _topic = topic;            
+            _clientId = clientId;
+
+            _maxBytesInOneMessageIncludingHeader = FetchRequest.DefaultBufferSize;
+        }
+
+        /// <summary>
+        /// Updating the cosumerGroup's offset for the partition in topic
+        /// </summary>
+        /// <param name="consumerGroup">The consumer group</param>
+        /// <param name="offset">The new offset. must be larger than or equal to zero</param>
+        /// <returns></returns>
+        public async Task UpdateOrCreateOffset(string consumerGroup, long offset)
+        {
+            if (string.IsNullOrEmpty(consumerGroup)) throw new ArgumentNullException("consumerGroup");
+            if (offset < 0) throw new ArgumentOutOfRangeException("offset", "offset must be positive or zero");
+
+            OffsetCommitRequest request = CreateOffsetCommitRequest(offset, consumerGroup);
+            await _gateway.SendProtocolRequest(request, _topic, _partitionId);                                
+        }
+
+        /// <summary>
+        /// Get the max offset of the partition in the topic.
+        /// </summary>
+        /// <returns>The max offset, if no such offset found then returns -1</returns>
+        public async Task<long> GetLastOffset()
+        {
+            var request = CreateGetLastOffsetRequest();
+            var response = await _gateway.SendProtocolRequest(request, _topic, _partitionId);
+            return response.Offsets.Count > 0 ? response.Offsets.First() : NoOffsetFound;
+        }
+
+        /// <summary>
+        /// Getting the offset of a specific consumer group
+        /// </summary>
+        /// <param name="consumerGroup">The name of the consumer group</param>
+        /// <returns>The current offset of the consumerGroup</returns>
+        public async Task<long> GetOffset(string consumerGroup)
+        {
+            if (string.IsNullOrEmpty(consumerGroup)) throw new ArgumentNullException("consumerGroup");
+
+            OffsetFetchRequest offsetFetchrequest = CreateOffsetFetchRequest(consumerGroup);
+
+            // TODO: Should also bring timeout?
```

Comment:
We decided that this is our consumer logic.

so he should run this code with task.whenAny and decide what to do.